### ISSUE TITLE
Unexpected behavior using server-mode in incremental branch

### ISSUE THREAD

**Issue by karoliineh** [goblint/analyzer#623]
In the master branch, reanalysing an unchanged file behaves the same as expected. In the incremental branch, the first analysis works fine, but the second analysis result is different from the first one. For example, analysing the [01-assert.c](tests/regression/00-sanity/01-assert.c) file, the first run in the incremental gives:

```
...
[Success][Assert] Assertion "success" will succeed (./src/individual.c:9:3-9:18)
[Error][Assert] Assertion "fail" will fail. (./src/individual.c:10:3-10:15)
[Warning][Integer > Overflow][CWE-191] Signed integer underflow (./src/individual.c:11:3-11:23)
[Warning][Assert] Assertion "unknown == 4" is unknown. (./src/individual.c:11:3-11:23)
...
```

but the second run gives an extra warning in the end:
```
...
[Warning][Unknown] Calculated state for undefined function: unexpected node Statement assert(silence);
```

The same happens with file [smtprc_comb.c](pthread/smtprc_comb.c). There the added warnings are:

```
...
[Warning][Unknown] Calculated state for undefined function: unexpected node Statement return (0);
[Warning][Unknown] Calculated state for undefined function: unexpected node Statement free((void *)buf);
[Warning][Unknown] Calculated state for undefined function: unexpected node Statement return (-1);
```

This behaviour occurs with `server.enabled` and `server.reparse` enabled.

Although this is an obvious unexpected behaviour I do have another question regarding this. The "new warnings" do not have a location. Right now I assume in GobPie that the **location is always provided by Goblint**. Are the missing locations directly related to this bug or is it possible, that Goblint does not give some information in some chances? The main question is if I should handle the "missing" case for each field or can I continue assuming that all information is always provided?

**Comment by sim642** [comment]
I haven't looked at this again, but some similar warnings we're fixed before here: https://github.com/goblint/analyzer/pull/591#issuecomment-1034909607. I guess there's some other data structure used on the `interactive` branch that still isn't reset correctly.

---

> The "new warnings" do not have a location. Right now I assume in GobPie that the **location is always provided by Goblint**. Are the missing locations directly related to this bug or is it possible, that Goblint does not give some information in some chances? The main question is if I should handle the "missing" case for each field or can I continue assuming that all information is always provided?

The missing locations are in some cases intentional and the `loc` field in message pieces is optional:
https://github.com/goblint/analyzer/blob/4c7576af93584aae20ce7a7ba6f3edc34a378fa7/src/util/messages.ml#L30-L36
So the missing case for that needs to be handled somehow, but there's no need to consider other fields as potentially missing. Depending on what Magpie allows, there are a few choices:
1. Completely ignore locationless messages.
2. If possible, output the message such that it's in the warnings list, but nowhere in the analyzed source code.
3. If the previous is not possible, then alternative might be to attach _some_ location to it to have it show up somewhere.

**Comment by sim642** [comment]
With some extra printing, the first run has results for:
```
result for node 19 "assert(silence);"
result for call of main
result for node 16 "assert(unknown == 4);"
result for node 11 "success = 1;"
result for node 13 "fail = 0;"
result for node 17 "return (0);"
result for entry state of main
result for node 14 "assert(success);"
result for node 15 "assert(fail);"
result for node 10037113751 "return;"
result for node 12 "silence = 1;"
```
And the second run has results for:
```
result for node 2 "silence = 1;"
result for node 19 "assert(silence);"
[Warning][Unknown] Calculated state for undefined function: unexpected node node 19 "assert(silence);"
result for entry state of main
result for node 5 "assert(fail);"
result for node 10137524190 "return;"
result for node 9 "assert(silence);"
result for node 7 "return (0);"
result for entry state of main
result for node 4 "assert(success);"
result for call of main
result for node 6 "assert(unknown == 4);"
result for node 10037113751 "return;"
result for node 1 "success = 1;"
result for node 3 "fail = 0;"
```

Despite the function being in `unchanged`, something happens with the comparison and updating such that three unknowns for the first run stay around:
```
result for entry state of main
result for node 19 "assert(silence);"
result for node 10037113751 "return;"
```

**Comment by michael-schwarz** [comment]
Hmm, could this also be related to the issue we are having with Gobview in general?

**Comment by sim642** [comment]
> Hmm, could this also be related to the issue we are having with Gobview in general?

I don't think so, because this particular issue only happens with `--enable server.reparse`, whereas Gobview doesn't do that. Also in Gobview all the statement IDs and nodes are mixed up, but here's it's just one statement really.

**Comment by sim642** [comment]
Something's also up with the function IDs because on the `interactive` branch, the identical reparsed file appears to have a different start variable:
```
Destabilizing sides of changed functions, primary old nodes and removed functions ...
Old start var entry state of main (305) on ./tests/regression/00-sanity/01-assert.c:4:1-14:1
Destabilizing and restarting removed start var entry state of main (596) on ./tests/regression/00-sanity/01-assert.c:4:1-14:1
```
